docs: plugin authoring workflow overview pages (#20847)
gatsbyjs/gatsby: docs/docs/loading-plugins-from-your-local-plugins-folder.md

title: Loading Plugins from Your Local Plugins Folder
---

Gatsby can load plugins from your website's local plugins folder, which is a folder named `plugins` in the website's root directory.

> Consider this example Gatsby website's project structure, which includes a local plugin called `gatsby-local-plugin`:

Reviewer: ```suggestion
Consider this example project structure which includes a local plugin called `gatsby-local-plugin`:
```

```
/my-gatsby-site
└── /src
    └── /pages
    └── /components
<!-- highlight-start -->
└── /plugins
    └── /gatsby-local-plugin
        └── /package.json
        └── /gatsby-node.js
<!-- highlight-end -->
└── gatsby-config.js
└── gatsby-node.js
└── package.json
```

Like the name of the plugins folder implies, you can include multiple plugins in your local plugin folder.

> Including a local plugin in your plugins folder also requires a configuration step (just like a 3rd party plugin you've installed into your `node_modules` by running `npm install`); just as plugins installed from npm need to be included in your `gatsby-config`, you need to add the name of your local plugin to the plugins array as well:

Reviewer: ```suggestion
Including a local plugin in your plugins folder also requires a configuration step (similar to a third-party plugin you've installed in your `node_modules` folder by running `npm install`); just as plugins installed from npm need to be included in your `gatsby-config`, you need to add the name of your local plugin to the plugins array as well:
```

```javascript:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-third-party-plugin`,
    `gatsby-local-plugin`, // highlight-line
  ],
}
```

## Verifying your plugin is loading

> To verify that your plugin is actually being found, you could add a small snippet of code like this to a `gatsby-node.js` file in the root of your plugin:

Reviewer: ```suggestion
To verify that your plugin is available for use in your Gatsby site, you can add a small snippet of code to a `gatsby-node.js` file in the root of your plugin:
```

Reviewer: Is it worth mentioning that you may not have a gatsby-node file in this directory already? Trying to think of the order of operations/use case here.

Author: Good point, I think that would be a good aside to say you may need to add one.

```javascript:title=plugins/gatsby-local-plugin/gatsby-node.js
exports.onPreInit = () => {
  console.log("Testing...")
}
```

_The [`onPreInit` API](/docs/node-apis/#onPreInit) is the first Node API called by Gatsby right after plugins are loaded._

> Then, when running your site in develop or build, you should see "Testing..." logged inside Gatsby's output in the console:

Reviewer: ```suggestion
Then, when running your site in develop or build mode, you should see "Testing..." logged in your terminal:
```
